### PR TITLE
Madison/context attn

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

**Comment by madisonmay** [comment]
Not yet ready for merge -- but would appreciate initial review

File: finetune/base.py [hunk 1]
```diff
@@ -145,10 +145,15 @@ def _target_model(
...
+            seq_mask = tf.sequence_mask(featurizer_state['lengths'])
             for key in ['features', 'explain_out']:
                 if key in featurizer_state:
+                    binary_mask = tf.constant(1.) - tf.cast(seq_mask, tf.float32)
+                    context_embed = context_embed * tf.expand_dims(binary_mask, -1)
+                    sum_context = tf.reduce_mean(context_embed, 1)
+                    mean_context = sum_context / tf.reduce_mean(binary_mask)
                     featurizer_state[key] = tf.concat(
-                        (featurizer_state[key], tf.reduce_mean(context_embed, 1)), -1
+                        (featurizer_state[key], mean_context), -1
```
Comment (lhz1029) [inline finetune/base.py]: good call dealing with length!

File: finetune/base_models/bert/model.py [hunk 2]
```diff
@@ -129,7 +129,7 @@ class RoBERTa(SourceModel):
         "lm_type": "mlm",
         "n_embed": 768,
         "n_epochs": 8,
-        "n_heads": 12,
+        "n_heads": 16,
```
Comment (lhz1029) [inline finetune/base_models/bert/model.py]: is this intentional?

File: finetune/base_models/gpt/featurizer.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ def mask_attn_weights(w):
 
 def mask_pad(w, lengths):
     batch = shape_list(lengths)[0]
-    maxlen = tf.maximum(lengths)
+    maxlen = tf.cast(tf.reduce_max(lengths), tf.int32)
```
Comment (lhz1029) [inline finetune/base_models/gpt/featurizer.py]: what's the difference in behavior? I'm having trouble learning about the differences online

File: finetune/encoding/sequence_encoder.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 import copy
 from collections import defaultdict
 
+import ipdb
```
Comment (lhz1029) [inline finetune/encoding/sequence_encoder.py]: probably can delete

File: finetune/nn/target_blocks.py [hunk 5]
```diff
@@ -356,6 +356,7 @@ def sequence_labeler(
     train=False,
     reuse=None,
     lengths=None,
+    context=None,
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py]: this doesn't seem to be used...is there a reason why we want to add this is?

File: finetune/config.py [hunk 6]
```diff
@@ -321,7 +321,7 @@ def get_default_config():
         max_masked_tokens=128,
 
         # Sequence Labeling
-        seq_num_heads=16,
+        # seq_num_heads=16,
```
Comment (benleetownsend) [inline finetune/config.py]: ???

File: finetune/model.py [hunk 7]
```diff
@@ -146,7 +146,9 @@ def _model_fn(features, labels, mode, params):
                 explain=build_explain,
             )
             if context is not None:
-                featurizer_state = embed_context(context, featurizer_state, params, train)
+                print("Embedding position")
```
Comment (benleetownsend) [inline finetune/model.py]: print

**Pull request closed without merging**